Instruction and data trace (refactor to use C++ classes)

andyduller opened the pull request:
Hi Andrew/Scott,
I've rewritten the original Siemens instruction and data trace patch so that it uses a C++ class. I have split it into 5 separate commits which hopefully makes things easier to review.
I think it is much simpler, and hopefully less invasive than it was before as it doesn't touch the sim_t class anymore, as suggested by Scott I think.

I believe there is still the question of the documentation of the CSV formats produced by the instruction and data trace; Gadge is looking at how best to provide this information to you.

Any questions, please let me know. Thanks,
Andy Duller

scottj97 commented:
Please clean up the git history: it should not have the original commit + revert, or the merge commit from upstream/master. (See any good blog on `git rebase --interactive`.) There's also a merge conflict with the latest master that will need to be resolved.

scottj97 commented:
I think this PR should explain a little more of the motivation here, especially regarding if/when this CSV format is intended to become standardized.

scottj97 commented:
Looking at the commits a little closer, this isn't really what I meant when I recommended breaking it up into pieces. It looks like you simply took the completed implementation and made a separate commit out of each final file.

Commit 0e93f2cdf7188bd66a36ffd3ee6ca259b6fd758c adds a complete new class that's unused and not even compiled. That's not useful.

Commit 0ad499a08c4b8d3f27fa6a330e85fe9e8ca31ce4 calls method `proc->set_i_trace()` that doesn't exist yet. Compilation fails.

What I meant is something like:
1. Add empty trace class and instantiate it inside `proc_t`
2. Add cmdline args and pass them into the tracer
3. Add basic trace of PC
4. Add tracing of load/stores
5. ...etc...

The commit history should be an archaeological record of how and (more importantly) why the code came to be the way it is. The commit messages are a place to record design decisions and rationale. That's not possible when an entire 326-line file suddenly appears in complete form.

I have lots of nits to pick on the code itself but we'll get to that in due time. If you don't want to go back and split this up properly it would be better to simply squash it all into one commit.

andyduller commented:
Sorry, just noticed all of these useful comments about yesterday's pull request so I'll remove today's attempt as it's clearly not what you wanted.
Thanks,
Andy

andyduller commented:
I'll discuss with Gadge what he wants to do next. Be in touch next week.

andyduller commented:
Hi Scott,
Regarding the archaeology of the code development, I'm afraid that just doesn't exist. It started as a simple patch against the isa-sim back in 2017 (I think) and has gone through a number of different revision control regimes since then.

I have now reverted to a single commit for the main code portions and rebased on the latest version of master. I've also added a README.processor_trace in that commit, to describe the CSV formats and to provide a reference to the source of most of the information regarding this work, namely:
https://github.com/riscv-non-isa/riscv-trace-spec

I've then added another commit with a set of tests against expected files. I'm afraid that these are the only type of tests that we have ever done to check that the spike output is as expected. What I have added is a small subset of the tests that are available in the riscv-trace-spec repository and more of them can be added if necessary. I have tried to keep the footprint of these small (I have compressed the expected files) and if you would like more added then perhaps we ought to use git-lfs?
Thanks,
Andy

scottj97 commented on README.processor_trace:
```diff
@@ -0,0 +1,100 @@
+
+https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/riscv-trace-spec.pdf
+
+In addition python refence models of the trace encoder and trace
```
reference is misspelled

andyduller replied:
Thanks

andyduller replied:
Corrected.

scottj97 commented:
I've only skimmed the trace spec. I understand there is Branch Trace and Data Trace, and those correspond to the two files output by this change.

Are these two trace streams truly independent, though? A real system could presumably correlate the two streams by the timing of the messages (though I couldn't find anything useful in the trace spec when I searched for "correlate").

I'm asking because I wonder if it would make more sense to output them all in one file, where each line was either a data or a branch trace.

andyduller commented:
Hi Scott,
Interesting thoughts. Since the two streams come from the same executing
program they are not independent in that sense. However, the instruction
trace will produce a trace for every executed instruction and the data
trace only outputs for a subset of instructions.
I think that when this "flavour" of instruction and data trace was
developed, by UltraSoC initially, it was envisaged that the two types of
trace were independent, since adding each type of trace would cost silicon
area. I believe that the expectation was that instruction trace would be
far more common than data trace.
Thanks,
Andy


On Fri, 11 Feb 2022 at 19:48, Scott Johnson ***@***.***>
wrote:

> I've only skimmed the trace spec. I understand there is Branch Trace and
> Data Trace, and those correspond to the two files output by this change.
>
> Are these two trace streams truly independent, though? A real system could
> presumably correlate the two streams by the timing of the messages (though
> I couldn't find anything useful in the trace spec when I searched for
> "correlate").
>
> I'm asking because I wonder if it would make more sense to output them all
> in one file, where each line was either a data or a branch trace.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv-software-src/riscv-isa-sim/pull/913#issuecomment-1036562357>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH3WEMGPWXYYE7OOEY77BDU2VRXZANCNFSM5M64NTYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

scottj97 commented on riscv/proc_trace.cc:
```diff
@@ -0,0 +1,325 @@
+        addr &= 0xffffffff;
+        tval &= 0xffffffff;
+    }
+    fprintf(m_itrace_fd, "1,%lx,%lx,%x,%x,%lx,%lx,%x\n",
```
This is not portable. If you must use C stdio, use the defines from `inttypes.h` such as PRIx64. You can see a few examples of this elsewhere in Spike.

scottj97 commented on riscv/proc_trace.cc:
```diff
@@ -0,0 +1,325 @@
+            // Not interested in load immediate
+            uint64_t inst_15_13 = (binary >> 13) & ((insn_bits_t(1) << 3) - 1);
+            uint64_t inst_1_0 = binary & 0x3;
+            m_is_load = (((inst_1_0 == 0) || (inst_1_0 == 2)) && (inst_15_13 > 0) && (inst_15_13 < 4));
+            m_is_store = ((inst_1_0 == 0) || (inst_1_0 == 2)) & (inst_15_13 > 4);
```
This is not robust, as new load & store opcodes might be added to the ISA at any time, or in a custom fork of Spike, and they would not know to update this cryptic decoder here.

Spike already has some way of decoding whether an opcode is a load/store, though it's not clear to me why this is needed here, since the act of loading or storing will call other methods in this class. That would be a much more reliable way of determining that an instruction did a load or store.

andyduller replied:
Could you please point me at the way of deciding whether an opcode is a load or store? I had looked in decode.h, but couldn't see any obvious way of doing it. I agree that I really don't want this partial disassembly in this code.
Thanks.

scottj97 replied:
What I meant is, if you see `record_load()`, `record_store()`, or `record_amo()` called, then you know it's a load/store. If not, then it's not.

andyduller replied:
I have removed the disassembly code. The check for whether it's an atomic instruction is required because otherwise you get spurious load/stores being recorded. I didn't have a test for this in the subset of tests I have imported, so I've added a test which requires this check.
I also need to be able to determine what type of instruction it is for the data trace of CSR and atomics so I've now done that using the constants defined in encoding.h.

scottj97 replied:
Can you explain what you mean by "spurious load/stores"? An AMO is both a load and a store. What's supposed to get recorded in the trace in that case? I wonder if a solution for #873 would be able to help here.

andyduller replied:
The data trace spec differentiates AMO from other load/store operations. The problem I am trying to solve here is, to make sure that the load/store operations are not recorded in the trace, if the instruction is actually an AMO.
Regarding #873, it may be that a solution to that helps here.

andyduller replied:
Table 4.12 in the trace spec (https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/riscv-trace-spec.pdf) gives a list of the types of data access that are to be recorded as part of the data trace.

scottj97 commented on riscv/proc_trace.cc:
```diff
@@ -0,0 +1,325 @@
+            // It's not clear why these are being triggered.
+            // They seem to come from DATA_SRC_LOAD_TLB_1
+            if (m_data_trace_debug)
+                fprintf(m_dtrace_fd, "NOT LOAD SRC %d %lx ld %d st %d amo %d %lx\n",
+                        src, addr, m_is_load, m_is_store, m_is_amo, m_insn_binary);
+            else
+                fprintf(stderr, "NOT LOAD SRC %d %lx ld %d st %d amo %d %lx\n",
+                        src, addr, m_is_load, m_is_store, m_is_amo, m_insn_binary);
```
No need to duplicate the entire` fprintf()` when the only difference is the output filehandle

andyduller replied:
Changed to using C++ streams.

scottj97 commented on riscv/proc_trace.cc:
```diff
@@ -0,0 +1,325 @@
+
+void proc_trace_t::record_csr_set_trace(int which, reg_t store_data) {
+    // mret is a special case where there is no call to csr_get to go with this.
+    // In this case the csr must be MSTATUS and it does not get recorded.
```
Why doesn't it get recorded? Is that specified in the trace spec?

andyduller replied:
I have removed the magic number in this function and replaced with a constant defined in encoding.h

scottj97 commented on spike_main/spike.cc:
```diff
@@ -55,6 +55,9 @@ static void help(int exit_code = 1)
   fprintf(stderr, "                        This flag can be used multiple times.\n");
   fprintf(stderr, "  --rbb-port=<port>     Listen on <port> for remote bitbang connection\n");
   fprintf(stderr, "  --dump-dts            Print device tree string and exit\n");
+  fprintf(stderr, "  --ust-trace=<file>    Write Instruction trace file\n");
```
What does "UST" mean? Why not `--inst-trace`?

andyduller replied:
UST is UltraSoC Trace, it's historic as the work on instruction was initially done by UltraSoC in 2017.

scottj97 replied:
So the switch is "UltraSoC Trace trace"? Why not switch this to `--inst-trace`? There's no other references to UltraSoC that I can see in this PR, so no reason to keep this historical name.

andyduller replied:
Corrected.

scottj97 commented on README.processor_trace:
```diff
@@ -0,0 +1,100 @@
+-------
+
+A small collection of tests are included which run spike to generate instruction and data trace
+file and then checks the output against expected files. Both 64-bit and 32-bit binaries are
```
Where did the expected files originate? Has there been any independent validation of their contents?

andyduller replied:
They were produced by a run through a version of spike. The validation that has been done is twofold:
- the traces have been run through various models of the trace encoder and decoder within Siemens and by their customers, and it has been checked that the original trace was reconstructable.
- for some of the smaller files the data has been scrutinized within Siemens and by their customers to check that it agrees with hardware models. In particular, Seagate has contributed a "torture" test example as part of their RISCV development work (https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/referenceFlow/tests/examples/br_j_asm_test/test.S).

scottj97 replied:
My concern is how we keep these traces up to date over time. As Spike changes, these traces might need to change too.

I guess it wouldn't be too hard to eyeball any differences and then manually copy the new traces over the golden traces (assuming the changes are intentional)? Could/should we document that process?

Should we run these tests as part of the GitHub CI? Otherwise people will submit PRs and not even realize they've broken the tracing.

andyduller replied:
I share your concern over using expected files for CI.
It's hard to quantify, but our experience with the instruction trace files, over a number of years (data trace was much more recent), is that the trace output has not changed as we have updated the patch file to keep it up to date with your upstream changes. Given the fixed elf file input, I would expect that there should be no reason for the trace to change.

Looking at  some of the core instruction trace code
```
      proc_trace->set_addr(state.pc);
      proc_trace->set_insn(insn, bits);
      proc_trace->set_priv(state.prv);
```
e.g. if the encoding or data type of state.prv were to change, then I would expect that there would be large numbers of differences against the expected files (or it might not even compile?).

Regarding documenting how to deal with CI failures:
- the traces include the start-up code that is part of spike; if that changes then in terms of content or location, then the expected files will need to be updated
- if the data types of e.g. internal state of processors was to be changed, then the C++ formatted output of the trace data may well need to be changed
- given that the elf file input to the tests is fixed (and is not subject to changes in the toolchain), I would expect that the instructions addresses will not change (since proxy kernel operation is not supported for trace)

andyduller replied:
I suppose in that case there would probably need to be a re-compilation of the source and the new output from running the instruction and data trace taken as the expected output. Are you saying that, in the case that you have cited, the old elf files would not run using an updated spike?

scottj97 replied:
In the case I'm imagining the old elf files would not run the same way anymore, and would not give good trace coverage anymore. Maybe this is far-fetched.

There is also the case where we discover a problem with the tracing, and we want to expand coverage to make this problem visible, which means modifying one of the elfs, or adding a new elf.

I want to avoid the provenance of these elf files being a mystery to anyone new coming into this project. "I made these by hand once and I'm assuming they will never need to be changed" is not generally a good plan.

andyduller replied:
I agree that lack of provenance is not a good plan to build in. I'll create something to provide some provenance.

andyduller replied:
I've added extra information about how to rebuild the various test binaries. In particular, I've added the script and C entry point code required to build the embench code for use with spike.

sifiverobert replied:
I was expecting, that SPIKE output itself can be used to check if encoder+decoder will produce desired PC sequence.
Spike patch is producing PC sequence (+ other details). If we extract 'PC-only flow', we can check if encoder+decoder are is producing same PC flow. Using such a technique ANY file/program which can be run by spike can be used to test trace. Similar technique can be used to validate data trace - trace should reconstruct all memory transactions (same addresses and data).

scottj97 reviewed (changes requested):
My comments are mostly small things that could be improved, but I'm a little skeptical of this entire PR. It feels very invasive. I also don't like having a long list of class attributes that get reset and reused for each line of trace output. It would be better to somehow accumulate the info as it becomes available into a new object for each line of trace. (Though it's not obvious to me how best to do that.) That would make it impossible to accidentally mix up two instructions' fields in one line of trace.

This PR includes several binary test programs. Where did these binaries come from? How can we recreate them? The ISA changes over time and Spike changes with it, and sometimes that can necessitate updating these kind of test programs. (To initialize a new CSR, for example.) Would it be better if their source code were included here, and your script ran the compiler/assembler first?

andyduller commented on README.processor_trace:
```diff
@@ -0,0 +1,100 @@
+
+EXCEPTION: Whether an exception occurred (%x format, uint8_t)
+
+ECAUSE: The exception value when an exception occurred (%lx format, reg_t)
```
Corrected.

andyduller commented on README.processor_trace:
```diff
@@ -0,0 +1,100 @@
+automatically uncompresses these before comparing with the generated traces.
+
+The output is placed in a sub-directory of the current working directory, called trace_regression.
+Each run of the script will remove any previously run tests.
```
Corrected.

andyduller commented on riscv/data_tracer.h:
```diff
@@ -0,0 +1,25 @@
+    }
+    void trace(uint64_t addr, size_t bytes, access_type type) {
+        (void)addr;
+        (void)type;
```
Corrected.

andyduller commented on riscv/data_tracer.h:
```diff
@@ -0,0 +1,25 @@
+{
+ public:
+    datatracer_t() {}
+    bool interested_in_range(uint64_t begin, uint64_t end, access_type type) {
```
Corrected.

andyduller commented on riscv/processor.cc:
```diff
@@ -1032,7 +1088,10 @@ reg_t processor_t::get_csr(int which, insn_t insn, bool write, bool peek)
     if (!peek)
       search->second->verify_permissions(insn, write);
-    return search->second->read();
+    reg_t old = search->second->read();
```
Corrected.

andyduller commented on riscv/proc_trace.h:
```diff
@@ -0,0 +1,99 @@
+    void record_store(reg_t addr, uint64_t val, size_t size, data_src_t src);
+    void record_amo(reg_t addr, uint64_t load_data, uint64_t store_data, size_t size);
+    void record_csr_get_trace(int which, insn_t insn, bool write, reg_t load_data, size_t size);
+    void record_csr_set_trace(int which, reg_t store_data);
```
Corrected.

andyduller commented on riscv/proc_trace.cc:
```diff
@@ -0,0 +1,325 @@
+}
+
+void proc_trace_t::open_i_trace(const char *instruction_trace_file_name) {
+    m_itrace_fd = fopen(instruction_trace_file_name, "w");
```
Changed to using C++ streams.

andyduller commented on riscv/mmu.h:
```diff
@@ -164,6 +169,7 @@ class mmu_t
         return misaligned_store(addr, val, sizeof(type##_t), xlate_flags); \
       reg_t vpn = addr >> PGSHIFT; \
       size_t size = sizeof(type##_t); \
+      if (proc) proc->get_proc_trace()->record_store(addr, val, size, DATA_SRC_STORE); \
```
Also changed, as suggested, thanks. Data trace now requires --enable-commitlog to function correctly.

scottj97 replied:
Is there [code like this](https://github.com/riscv-software-src/riscv-isa-sim/blob/5ba63a44706173b556b8d5632872b39a09d7f16d/riscv/sim.cc#L250-L256) that will give a reasonable error message if someone forgets the configure option and asks for a data trace?

scottj97 replied:
It looks like you could also easily make `READ_MEM()` do the right thing here even if `RISCV_ENABLE_COMMITLOG` is disabled. The only reason to make it conditional on that compilation flag is if you can provide a performance improvement by disabling it.

andyduller replied:
Yes, I did wonder about whether to put the record_load/record_store in the #ifndef RISCV_ENABLE_COMMITLOG, but as you say, it gives a performance gain if data trace isn't being used.
I've added this check, a good idea, thanks.

andyduller commented on riscv/processor.cc:
```diff
@@ -999,6 +1041,18 @@ void processor_t::disasm(insn_t insn)
+      proc_trace->step();
+      // Need to translate addresses in a trap so do an mmu translate in that case.
+      // However, in general we want virtual addresses for the instruction trace.
+      if (state.pc & 0x7000000000000000ULL)
```
As you say, this is a strange magic number. To be honest, I can't remember where it came from.
The addresses recorded in the instruction trace need to be used by the trace decoder to look-up the instruction in the original elf file. I've made a change to this check, but I'm not sure it's the best way of expressing what is required. What do you think?

scottj97 replied:
You changed this to: `if (state.pc & (1ULL << MAX_PADDR_BITS))`

Which is different, but also unexplained. It seems the goal here is to record the physical address? And this is a (broken) attempt to determine if the PC value is a physical address? But the old comment says the trace wants virtual addresses. So I don't understand why you would ever call `mmu->translate()` here.

The new comment says "the decoder must be able to relate the pc addresses back to the elf file" -- but ELF files include both virtual addresses and physical addresses, and Spike will initialize memory based on the physical address. Some bare-metal test cases will lie to the assembler and use a physical address in place of the virtual, so I understand why sometimes the decoder will not be able to find a given virtual address in the ELF. But that's a problem better solved either upstream (when the ELF is created) or downstream (in the decoder). I don't see any reason to ever translate the address here.

andyduller replied:
Hmmm, ok, I don't know how to change this in that case.

scottj97 replied:
IIUC the instruction trace is supposed to record only virtual addresses, yes? The address should never be translated here.

If that causes problems elsewhere in the flow, those problems should be solved outside of Spike.

sifiverobert commented on riscv/execute.cc:
```diff
@@ -215,7 +215,7 @@ static inline reg_t execute_insn(processor_t* p, reg_t pc, insn_fetch_t fetch)
 bool processor_t::slow_path()
 {
-  return debug || state.single_step != state.STEP_NONE || state.debug_mode;
+  return i_trace | d_trace | debug || state.single_step != state.STEP_NONE || state.debug_mode;
```
This '|' instead of '||' seems like logical bug (but it will work). Below it is correct.

sifiverobert commented on tests/processor_trace/build_files/README.embench:
```diff
@@ -0,0 +1,18 @@
+within the embench-iot repository.
+
+Note that the build script requires access to a 32-bit RISCV compiler. The embench_build.sh needs to be modified to be able
+to locate this compiler.
```
What about 64-bit testing? Here it says '32-bit' only ...

sifiverobert commented:
Hi Andy & Scott,
I accidentally stepped into this pull-request and commented a bit. I hope you do not mind.
I am interested in this as I am a chair of Nexus Trace TG and I want to connect to Spike output as well. Some initial verification infrastructure (for Nexus Trace ...) is available here: https://github.com/riscv-non-isa/tg-nexus-trace/blob/master/refcode/c/NexusTraceTG-RefCode.pdf. I was using something what I called 'PCSEQ' file, but I want to use Spike output instead.
